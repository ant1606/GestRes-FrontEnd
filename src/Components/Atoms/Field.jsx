import React from 'react'
import 'animate.css';

const Field = ({ type, label, name, value='', classBox = '', handleChange = null, errorInput = false }) => {
  //TODO: Verificar que el mensaje de error se agregue en esta seccion misma
  return (
    <div className={`${errorInput ? "animate__animated animate__headShake" : ""} relative ${classBox}`}>
      <input
        type={type}
        id={name}
        placeholder=" "
        name={name}
        className={`${errorInput ? "border-2 border-rose-500 text-rose-500 " : "border-gray-900"}
        px-3 py-1  border-b-2 bg-white peer w-full text-base transition-colors h-8
        focus:outline-none`}
        onChange={handleChange}
        value={value}
        autoComplete="off"
      />
      <label htmlFor={name}
        className={`${errorInput ? "text-rose-500 peer-focus:text-rose-500" : "text-gray-600 peer-focus:text-gray-900"} 
        absolute left-0 -top-4 text-xs cursor-text transition-all px-0        
        peer-placeholder-shown:top-1 peer-placeholder-shown:text-base peer-placeholder-shown:px-3
        peer-focus:-top-4 peer-focus:text-xs  peer-focus:px-0`}
      >
        {label}
      </label>
    </div>
  )
}

export default Field