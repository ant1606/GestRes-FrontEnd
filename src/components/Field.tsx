import React from 'react';
import 'animate.css';

interface Props {
  type: string;
  label: string;
  name: string;
  value: string | number | readonly string[] | undefined;
  classBox: string;
  handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  errorInput: string | null;
  min?: string;
  disabled?: boolean;
}

const Field: React.FC<Props> = ({
  type,
  label,
  name,
  value,
  classBox,
  handleChange,
  errorInput,
  ...props
}) => {
  const validateLengthErrorInput = errorInput !== null ? errorInput.trim().length : 0;
  return (
    <>
      <div className={`flex flex-col relative ${classBox}`}>
        <div
          className={`${validateLengthErrorInput > 0 ? 'animate__animated animate__headShake' : ''} 
        relative `}>
          <input
            type={type}
            id={name}
            placeholder=" "
            name={name}
            // eslint-disable-next-line prettier/prettier
            className={`${validateLengthErrorInput > 0 ? 'border-2 border-rose-500 text-rose-500 ' : 'border-gray-900'}
            px-3 py-1  border-b-2 bg-white peer w-full text-base transition-colors h-8 focus:outline-none`}
            onChange={handleChange}
            value={value}
            autoComplete="off"
            data-testid={name}
            {...props}
          />
          <label
            htmlFor={name}
            // eslint-disable-next-line prettier/prettier
            className={`${validateLengthErrorInput > 0 ? 'text-rose-500 peer-focus:text-rose-500' : 'text-gray-600 peer-focus:text-gray-900'} 
          absolute left-0 -top-4 text-xs cursor-text transition-all px-0 peer-placeholder-shown:top-1 peer-placeholder-shown:text-base peer-placeholder-shown:px-3
          peer-focus:-top-4 peer-focus:text-xs  peer-focus:px-0`}>
            {label}
          </label>
        </div>
        {validateLengthErrorInput > 0 && (
          <span className="text-xs absolute -bottom-5 z-10 text-red-500 font-bold ">
            {errorInput}
          </span>
        )}
      </div>
    </>
  );
};

export default Field;
