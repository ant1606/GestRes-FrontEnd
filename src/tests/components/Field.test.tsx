import { fireEvent, render } from '@testing-library/react';
import Field from '#/components/Field';
import { vi } from 'vitest';

describe('Componente Field', () => {
  const events = {
    handleChange(
      evt:
        | React.ChangeEvent<HTMLInputElement>
        | React.ChangeEvent<HTMLSelectElement>
        | React.ChangeEvent<HTMLTextAreaElement>
    ) {
      return evt;
    }
  };

  test('Debe renderizarse sin error', () => {
    const wrapper = render(
      <Field
        type="text"
        label="Name"
        name="name"
        value=""
        classBox=""
        handleChange={events.handleChange}
        errorInput={null}
      />
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  test('Se muestra el label correctamente', () => {
    const { getByLabelText } = render(
      <Field
        type="text"
        label="Name"
        name="name"
        value=""
        classBox=""
        handleChange={events.handleChange}
        errorInput={null}
      />
    );
    expect(getByLabelText(/Name/i)).toBeInTheDocument();
  });

  test('Debe mostrarse el mensaje de error', () => {
    const wrapper = render(
      <Field
        type="text"
        label="Name"
        name="name"
        value=""
        classBox=""
        handleChange={events.handleChange}
        errorInput="Error message"
      />
    );

    expect(wrapper.getByText(/Error message/i)).toBeInTheDocument();
  });

  test('Debe llamar al evento handleChange al momento de ingresar un valor al campo', () => {
    const handleChangeSpy = vi.spyOn(events, 'handleChange');
    const wrapper = render(
      <Field
        type="text"
        label="Name"
        name="name"
        value=""
        classBox=""
        handleChange={events.handleChange}
        errorInput={null}
      />
    );
    expect(handleChangeSpy).not.toHaveBeenCalled();
    fireEvent.change(wrapper.getByTestId('name'), { target: { value: 'John' } });
    expect(handleChangeSpy).toHaveBeenCalled();
  });

  test('Debe mostrar el valor inicial correctamente', () => {
    const { getByTestId } = render(
      <Field
        type="text"
        label="Name"
        name="name"
        value="John"
        classBox=""
        handleChange={events.handleChange}
        errorInput={null}
      />
    );
    expect(getByTestId('name').nodeValue).toBe('John');
  });

  test('Debe aplicar la clase de error al campo de entrada cuando hay un error', () => {
    const { getByTestId } = render(
      <Field
        type="text"
        label="Name"
        name="name"
        value=""
        classBox=""
        handleChange={events.handleChange}
        errorInput="Error message"
      />
    );
    expect(getByTestId('name')).toHaveClass('border-2 border-rose-500 text-rose-500');
  });

  test('No debe aplicar la clase de error al campo de entrada cuando no hay un error', () => {
    const { getByTestId } = render(
      <Field
        type="text"
        label="Name"
        name="name"
        value=""
        classBox=""
        handleChange={events.handleChange}
        errorInput={null}
      />
    );
    expect(getByTestId('name')).not.toHaveClass('border-2 border-rose-500 text-rose-500');
  });
});
