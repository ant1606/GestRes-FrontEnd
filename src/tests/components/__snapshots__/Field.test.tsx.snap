// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente Field > Debe renderizarse sin error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col relative "
      >
        <div
          class=" 
        relative "
        >
          <input
            autocomplete="off"
            class="border-gray-900
            px-3 py-1  border-b-2 bg-white peer w-full text-base transition-colors h-8 focus:outline-none "
            data-testid="name"
            id="name"
            name="name"
            placeholder=" "
            type="text"
            value=""
          />
          <label
            class="text-gray-600 peer-focus:text-gray-900 
          absolute left-0 -top-4 text-xs cursor-text transition-all px-0 peer-placeholder-shown:top-1 peer-placeholder-shown:text-base peer-placeholder-shown:px-3
          peer-focus:-top-4 peer-focus:text-xs  peer-focus:px-0"
            for="name"
          >
            Name
          </label>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col relative "
    >
      <div
        class=" 
        relative "
      >
        <input
          autocomplete="off"
          class="border-gray-900
            px-3 py-1  border-b-2 bg-white peer w-full text-base transition-colors h-8 focus:outline-none "
          data-testid="name"
          id="name"
          name="name"
          placeholder=" "
          type="text"
          value=""
        />
        <label
          class="text-gray-600 peer-focus:text-gray-900 
          absolute left-0 -top-4 text-xs cursor-text transition-all px-0 peer-placeholder-shown:top-1 peer-placeholder-shown:text-base peer-placeholder-shown:px-3
          peer-focus:-top-4 peer-focus:text-xs  peer-focus:px-0"
          for="name"
        >
          Name
        </label>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
